ui/src/components/credentials/CredentialLinkQR.tsx

```diff
@@ -0,0 +1,308 @@
+          if (response.data.status !== "pending") {
+            setCredentialQRCheck({ data: response.data, status: "successful" });
+
+            const { proofTypes } = shareCredentialQRCode.data.linkDetail;
+
+            if (proofTypes.includes("SparseMerkleTreeProof")) {
+              void message.info("Issuance process started");
+            }
+
+            if (proofTypes.includes("BJJSignature2021")) {
+              void message.success("Credential sent");
+            }
+          }
+        } else {
+          setCredentialQRCheck({ error: response.error, status: "failed" });
+
+          void message.error(response.error.message);
+        }
+      }
+    };
+
+    const checkQRCredentialStatusTimer = setInterval(() => {
+      if (
+        (isAsyncTaskDataAvailable(credentialQRCheck) &&
+          credentialQRCheck.data.status !== "pending") ||
+        hasAsyncTaskFailed(credentialQRCheck)
+      ) {
+        clearInterval(checkQRCredentialStatusTimer);
+      } else {
+        void checkCredentialQRCode();
+      }
+    }, POLLING_INTERVAL);
+
+    return () => clearInterval(checkQRCredentialStatusTimer);
+  }, [shareCredentialQRCode, linkID, credentialQRCheck, env]);
+
+  const onStartAgain = () => {
+    makeRequestAbortable(createCredentialQR);
+    setCredentialQRCheck({ status: "pending" });
+  };
+
+  const hasFailed = hasAsyncTaskFailed(shareCredentialQRCode)
+    ? shareCredentialQRCode.error
+    : hasAsyncTaskFailed(credentialQRCheck)
+    ? credentialQRCheck.error
+    : undefined;
+
+  if (hasFailed) {
+    if (hasFailed.status === 404) {
+      return (
+        <Space align="center" direction="vertical" size="large">
+          <Avatar className="avatar-color-error" icon={<QRIcon />} size={56} />
+
+          <Typography.Title level={2}>Credential link is invalid</Typography.Title>
+
+          <Typography.Text type="secondary">
+            In case you think this is an error, please contact the issuer of this claim.
+          </Typography.Text>
+        </Space>
+      );
+    } else if (hasFailed.status === 400) {
+      return (
+        <Space align="center" direction="vertical" size="large">
+          <Avatar className="avatar-color-error" icon={<AlertIcon />} size={56} />
+
+          <Typography.Title level={2}>
+            The credential link has expired, please start again
+          </Typography.Title>
+
+          <Button icon={<IconRefresh />} onClick={onStartAgain} type="link">
+            Start again
+          </Button>
+        </Space>
+      );
+    }
+    return (
+      <ErrorResult error={hasFailed.message} labelRetry="Start again" onRetry={onStartAgain} />
+    );
+  }
+
+  if (!isAsyncTaskDataAvailable(shareCredentialQRCode)) {
+    return <LoadingResult />;
+  }
+
+  if (isAsyncTaskDataAvailable(credentialQRCheck) && credentialQRCheck.data.status !== "pending") {
+    const { proofTypes } = shareCredentialQRCode.data.linkDetail;
+    if (proofTypes.length > 1) {
```

Comment:
I'd be more explicit here and use:

`proofTypes.includes("SparseMerkleTreeProof") && proofTypes.includes("BJJSignature2021")`

Reply:
Indeed all this "includes" way of checking is a sign of fragile typing and parsing. We should be using a union of tuples here that leaves only the limited amount of possibilities we've to manage, without the need for the last fallback case.